pkg/wrangler/context.go

```diff
@@ -0,0 +1,304 @@
+	RESTMapper              meta.RESTMapper
+	SharedControllerFactory controller.SharedControllerFactory
+	leadership              *leader.Manager
+	controllerLock          *sync.Mutex
+
+	RESTClientGetter genericclioptions.RESTClientGetter
+
+	mgmt *management.Factory
+	core *core.Factory
+
+	started bool
+}
+
+type MultiClusterManager interface {
+	NormanSchemas() *types.Schemas
+	ClusterDialer(clusterID string) func(ctx context.Context, network, address string) (net.Conn, error)
+	Start(ctx context.Context) error
+	Wait(ctx context.Context)
+	Middleware(next http.Handler) http.Handler
+	K8sClient(clusterName string) (kubernetes.Interface, error)
+}
+
+// OnLeader registers a callback function to be executed when the current context becomes the leader.
+// It calls the OnLeader method of the leadership field in the Context struct and passes the provided function as an argument.
+func (w *Context) OnLeader(f func(ctx context.Context) error) {
+	w.leadership.OnLeader(f)
+}
+
+// StartWithTransaction is a method of the Context struct that starts a transaction and executes a provided callback function within the transaction. It returns an error if the callback fails
+func (w *Context) StartWithTransaction(ctx context.Context, f func(context.Context) error) (err error) {
+	transaction := controller.NewHandlerTransaction(ctx)
+	if err := f(transaction); err != nil {
+		transaction.Rollback()
+		return err
+	}
+
+	if err := w.ControllerFactory.SharedCacheFactory().Start(ctx); err != nil {
+		transaction.Rollback()
+		return err
+	}
+
+	w.ControllerFactory.SharedCacheFactory().WaitForCacheSync(ctx)
+	transaction.Commit()
+	return w.Start(ctx)
+}
+
+// Start registers the current context as started and performs necessary initialization steps.
+// It acquires a lock using the controllerLock mutex to ensure thread safety.
+// If the context is not already started, it calls the Register method of the Dynamic field in the Context struct,
+// passing the provided context and SharedControllerFactory as arguments.
+// It sets the started field to true.
+// It then calls the Start method of the ControllerFactory field, passing the provided context and 50 as arguments.
+// It returns an error if there is any.
+// It finally calls the Start method of the leadership field in the Context struct, passing the provided context.
+func (w *Context) Start(ctx context.Context) error {
+	w.controllerLock.Lock()
+	defer w.controllerLock.Unlock()
+
+	if !w.started {
+		if err := w.Dynamic.Register(ctx, w.SharedControllerFactory); err != nil {
+			return err
+		}
+		w.started = true
+	}
+
+	if err := w.ControllerFactory.Start(ctx, 50); err != nil {
+		return err
+	}
+	w.leadership.Start(ctx)
+	return nil
+}
+
+// WithAgent returns a shallow copy of the Context that has been configured to use a user agent in its
+// clients that is the given userAgent appended to "rancher-%s-%s".
+func (w *Context) WithAgent(userAgent string) *Context {
```

Comment:
Do we need this func? Is this something we can add later? That way we can remove the settings file if we don't need it.

Reply:
it's not a deal breaker just a question, otherwise it's an approved.

Reply:
You're right. It's not needed, and I removed it.